contrib/src/main/scala/akka/stream/contrib/SourceGen.scala

```
@@ -0,0 +1,120 @@
+
+    val fanOut2Stage = fanOut2unfoldingStage[O, S, E](seed, {
+      (handleStateElement, grabIn, completeStage) =>
+        new InHandler {
+          override def onPush() = {
+            val o = grabIn()
+            unfoldWith(o) match {
+              case None         => completeStage()
+              case Some((s, e)) => handleStateElement(s, e)
+            }
+          }
+        }
+    })
+
+    unfoldFlowGraph(fanOut2Stage, flow)
+  }
+
+  private[akka] def unfoldFlowGraph[E, S, O, M](
+    fanOut2Stage: GraphStage[FanOutShape2[O, S, E]],
+    flow:         Graph[FlowShape[S, O], M]
+  ): Source[E, M] = Source.fromGraph(GraphDSL.create(flow) {
+    implicit b =>
+      {
+        f =>
+          {
+            import GraphDSL.Implicits._
+
+            val fo2 = b.add(fanOut2Stage)
+            fo2.out0 ~> f ~> fo2.in
+            SourceShape(fo2.out1)
+          }
+      }
+  })
+
+  private[akka] def fanOut2unfoldingStage[O, S, E](seed: S, withInHandler: ((S, E) => Unit, () => O, () => Unit) => InHandler) = new GraphStage[FanOutShape2[O, S, E]] {
+
+    override val shape = new FanOutShape2[O, S, E]("unfoldFlow")
+    override def createLogic(attributes: Attributes) = {
+
+      new GraphStageLogic(shape) {
+
+        import shape._
+
+        var pending: S = seed
+        var pushedToCycle = false
+
+        setHandler(in, withInHandler((s, e) => {
+          pending = s
+          push(out1, e)
+          pushedToCycle = false
+        }, () => grab(in), completeStage))
+
+        setHandler(out0, new OutHandler {
+          override def onPull() = if (!pushedToCycle && isAvailable(out1)) {
+            push(out0, pending)
+            pending = null.asInstanceOf[S]
+            pushedToCycle = true
+          }
+
+          override def onDownstreamFinish() = {
+            //Do Nothing, intercept completion as downstream
```

close, i.e complete gracefully.
so if inner flow was aborted (due to some exception), and I happen to get `feedback.onDownstreamFinish` before `nextElem.onUpstreamFailure(error)`, this will swallow the error.